### PR TITLE
JENKINS-43079# Trusted release related changes

### PR DESCRIPTION
**Pull request by vivek**
# Description

See [JENKINS-43079](https://issues.jenkins-ci.org/browse/JENKINS-43079).

How this works:
- Release build is run in a docker container with Java8 and maven 3.3.9
- Works with user scoped credentials
- Create github credentials ( github username and accessToken) and Jenkins credentials (username and password) in **user store**
- Chose which user is doing release from the drop down. At present there are only few, we can add more as needed

For it to work we need changes in pom file where it must be `https://` url under scm developerConnection setting. _This means anyone releasing from their laptop (which we do not want anyways) will no longer be able to use their ssh keys. However there are ways to automate it using GitHub access token._

I have placed release pipeline script under `release` folder for the time being mostly to track changes, we should probably move it out in it's own repo.

Tested doing beta release of another (smaller) plugin on dogfood. It needs to be tested on release environment whenever that gets fixed. 

Script uses credential input step that are not supported yet in BlueOcean, here is how the build screen looks like to do release. 

image

image

TODOs:

[ ] Test it in proper env to release beta version of blueocean

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by michaelneale** [comment]
sounds like it might be ok @vivek 
Not sure where we will do a 1.1 from though - so up to you when/if this merges.

**Comment by vivek** [comment]
@michaelneale yeah, I guess I would like  to merge it but will have to wait for actually releasing using it till env is ready. 

In any case, we want to do a beta release first from release env (to avoid any unexpected error that might have been missed) and then actually do 1.1 or 1.2 release based on how soon we have the env. Sounds good?

**Comment by michaelneale** [comment]
@vivek yes - not sure we will have the release environment in time though.

File: release/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                        sh 'git push -u origin $BRANCH'
+                    }
+
+                    //Sanity build and test
+                    stage('Build and test') {
+                        sh 'echo Java version: `java -version`'
+                        sh 'echo Maven version: `mvn --version`'
+                        sh 'echo Git version: `git --version`'
+
+                        sh 'mvn -T 1C  clean install'
```
Comment (michaelneale) [inline release/Jenkinsfile]: I changed my mind - we can drop the test. We want to trust the ATH and previous pipeline runs here anyway.

File: release/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    deleteDir()
+
+    docker.image('maven:3.3-jdk-8').inside() {
+        def gitNameEmail = (env.GIT_AUTHOR_AND_GIT_EMAIL).split(':')
+        println gitNameEmail[0] + ":" + gitNameEmail[1]
+        String gitName = gitNameEmail[0]
+        String gitEmail = gitNameEmail[1]
+        withEnv(["GIT_COMMITTER_EMAIL=$gitEmail", "GIT_COMMITTER_NAME=$gitName", "GIT_AUTHOR_NAME=$gitName", "GIT_AUTHOR_EMAIL=$gitEmail"]) {
+            withCredentials([
+                    usernamePassword(credentialsId: '${JENKINS_CREDENTIALS}', passwordVariable: 'JENKINS_PASSWORD', usernameVariable: 'JENKINS_USERNAME'),
```
Comment (michaelneale) [inline release/Jenkinsfile]: should this be the cleartext version of it or as encrypted in my desktop settings.xml? 
(probably need to go over this some more)

Comment (vivek) [inline release/Jenkinsfile]: Its clear text. If it's running in insecure env, we better keep them encrypted but then we need to tell maven where encryption keys are so anyone who can get to env variable can possibly get to those encryption keys. Something we should run it by Ben?

**Comment by michaelneale** [comment]
my version of the Jenkinsfile: 

https://gist.github.com/michaelneale/821e20c6d417b9ae89eea91599d00183

**Comment by vivek** [comment]
Merging it in order to test a release, main thing is developerConnection url. Once a script is tested to work, we will update it in separate PR.

**Pull request merged**